agent/structs/config_entry_gateways.go

```diff
@@ -539,6 +574,56 @@ func (e *TerminatingGatewayConfigEntry) Validate() error {
 			return fmt.Errorf("Service %q must have a CertFile, CAFile, and KeyFile specified for TLS origination", svc.Name)
 		}
 	}
+
+	endpointSet := make(map[string]bool)
+
+	for _, endpoint := range e.Endpoints {
+		if endpoint.Name == "" {
+			return fmt.Errorf("Endpoint name cannot be blank.")
+		}
+
+		// Check for duplicate names
+		if ok := endpointSet[endpoint.Name]; ok {
+			return fmt.Errorf("Endpoint %s was specified more than once in the gateway", endpoint.Name)
+		}
+		endpointSet[endpoint.Name] = true
+
+		if err := validateLinkedEndpointAddress(endpoint.Address); err != nil {
+			return fmt.Errorf("Endpoint Name is invalid %w", err)
+		}
+
+		if endpoint.Port < 1 || endpoint.Port > 65535 {
+			return fmt.Errorf("Invalid Port number %d", endpoint.Port)
+		}
+
+		// If either client cert config file was specified then the CA file, client cert, and key file must be specified
+		// Specifying only a CAFile is allowed for one-way TLS
+		if (endpoint.CertFile != "" || endpoint.KeyFile != "") &&
+			!(endpoint.CAFile != "" && endpoint.CertFile != "" && endpoint.KeyFile != "") {
+
+			return fmt.Errorf("Endpoint %q must have a CertFile, CAFile, and KeyFile specified for TLS origination", endpoint.Name)
+		}
+	}
+
+	return nil
+}
+
+func validateLinkedEndpointAddress(address string) error {
+	var valid bool
+
+	ip := net.ParseIP(address)
+	valid = ip != nil
+
+	_, _, err := net.ParseCIDR(address)
+	valid = valid || err == nil
+
+	// Hostname
+	re := regexp.MustCompile(DNSName)
```

Comment:
Maybe for consistency we should use something like [validateHost](https://github.com/hashicorp/consul/blob/c6ff4ba7d861a9ea1ec875141e2200ad0faa488e/agent/structs/config_entry_gateways.go#L343-L343)

Reply:
yes, envoy support wildcard as a subdomain only. so *.domain.com would work but api.*main.com won't, see [here](https://github.com/envoyproxy/envoy/blob/ad667a2c71c2081236368dad450abb879afc4fb8/api/envoy/config/listener/v3/listener_components.proto#L156)

Reply:
I changed it to `validateHost`, but had to get rid of a lot of test cases for invalid IPs and CIDRs because it lets pretty much anything slide.